### DISCUSSION TITLE
code_browser.py importError: cannot import name 'VerticalScroll' from 'textual.containers'

### DISCUSSION THREAD

**Discussion by gaslitbytech** [Textualize/textual#2068] [Q&A] [answered]
Following [getting started](https://textual.textualize.io/getting_started/) when running in a clean 3.11 venv on an intel mac:

```sh
cd textual/examples/
python code_browser.py ../
```

I get:

```sh
Traceback (most recent call last):
  File "/Users/david/Community/textual/examples/code_browser.py", line 17, in <module>
    from textual.containers import Container, VerticalScroll
ImportError: cannot import name 'VerticalScroll' from 'textual.containers' (/Users/david/Community/textual/examples/.venv/lib/python3.11/site-packages/textual/containers.py)
```

```sh
# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.15.1 |
| Rich    | 13.3.2 |

## Python

| Name           | Value                                                        |
|----------------|--------------------------------------------------------------|
| Version        | 3.11.2                                                       |
| Implementation | CPython                                                      |
| Compiler       | Clang 13.0.0 (clang-1300.0.29.30)                            |
| Executable     | /Users/david/Community/textual/examples/.venv/bin/python3.11 |

## Operating System

| Name    | Value                                                                                             |
|---------|---------------------------------------------------------------------------------------------------|
| System  | Darwin                                                                                            |
| Release | 21.6.0                                                                                            |
| Version | Darwin Kernel Version 21.6.0: Mon Dec 19 20:44:01 PST 2022; root:xnu-8020.240.18~2/RELEASE_X86_64 |

## Terminal

| Name                 | Value           |
|----------------------|-----------------|
| Terminal Application | vscode (1.76.1) |
| TERM                 | xterm-256color  |
| COLORTERM            | truecolor       |
| FORCE_COLOR          | *Not set*       |
| NO_COLOR             | *Not set*       |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=227, height=25 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 227                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 25                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |
```

If I replace `VerticalScroll` with `Vertical` in `examples/code_browser.py` it works.

**Comment by davep** [comment, marked as answer]
From what you're saying, you seem to be using the latest release of Textual from PyPi to run the examples from `main` in the repo, is that correct?

Some breaking changes are currently happening so this would be expected if you were to mix these. If you are checking out the examples it'll be better to check out those [tagged v0.15.1](https://github.com/Textualize/textual/tree/v0.15.1).

**Comment by gaslitbytech** [comment]
Thanks that does the trick. I initially discounted that as it was released only yesterday!

Perhaps the docs when they suggest:

```sh
git clone https://github.com/Textualize/textual.git
```

To checkout the most recent release tag.

**Reply by davep** [reply]
> I initially discounted that as it was released only yesterday!

We do try and keep busy. ;-)

Good point on the clone to get the examples in the docs; we'll have a think on that. Although explaining to folk how to get the correct tag might be more work than the occasional need to answer this question. But we'll take a look, thanks.

**Reply by willmcgugan** [reply]
Is there a way to link to the most recent tag? Otherwise we would have to update that link every release.

**Reply by davep** [reply]
A quick search suggests you can't link to the latest tag of the repo as a URL you can clone (`latest` is a valid target for the release URL -- [https://github.com/Textualize/textual/releases/latest](https://github.com/Textualize/textual/releases/latest) -- but that does't help here).

**Reply by gaslitbytech** [reply]
I was thinking with the top right:

<img width="197" alt="Screen Shot 2023-03-15 at 9 51 26 pm" src="https://user-images.githubusercontent.com/829044/225314186-4fd45893-df09-4f91-9d2f-374cfeffd31c.png">

It may be possible. 

I see that this is done with `mkdocs-material` from View Source:

```html
<li class="md-source__fact md-source__fact--version">v0.15.1</li>
```

Could this be used to dynamically change the code snippet and just add the following after the clone?

```sh
git checkout v0.15.1
```

**Reply by gaslitbytech** [reply]
From the [mkdocs-material code-blocks docs](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/) I've peiced the following together though unsure how to test it.

```patch
diff --git a/docs/getting_started.md b/docs/getting_started.md
index 82739d33..a4f0e0be 100644
--- a/docs/getting_started.md
+++ b/docs/getting_started.md
@@ -71,13 +71,14 @@ The Textual repository comes with a number of example apps. To try out the examp
     ```
 
 
-With the repository cloned, navigate to the `/examples/` directory where you will find a number of Python files you can run from the command line:
+With the repository cloned, checkout latest release:
 
-```bash
-cd textual/examples/
-python code_browser.py ../
+``` { .bash .annotate }
+git checkout "data-md-source__fact--version"
 +```
 
+And now navigate to the `/examples/` directory where you will find a number of Python files you can run from the command line:
+
 
 ## Textual CLI
 
```

**Reply by davep** [reply]
At first glance I'm not sure I quite follow what's going on there (which is fine, I'm just looking in passing), but one potential issue I see here is that if this is relying on the version information that is gathered from our repo on GitHub, this likely isn't going to play well with the offline mode of building the docs (which is important to some Linux distributions -- they need that the docs can be built without an Internet connection and, as such, the plugin that handles the GitHub badge is never included).

**Comment by gaslitbytech** [comment]
> with the offline mode of building the docs (which is important to some Linux distributions -- they need that the docs can be built without an Internet connection and, as such, the plugin that handles the GitHub badge is never included).

This is insightful thanks @davep. I'll close this discussion again now. Thanks for the help.

**Reply by gaslitbytech** [reply]
I'm just temporally reopening this again for a follow up possible workable solution.

> with the offline mode of building the docs (which is important to some Linux distributions -- they need that the docs can be built without an Internet connection and, as such, the plugin that handles the GitHub badge is never included).

It involves `poetry` that I'm not much familiar with. Will the offline mode have access to `version = "0.15.1"` thats baked in the `pyproject.toml`. Could this be used to build the docs without an internet connection?
